lib/utils/compose.coffee

```diff
@@ -0,0 +1,747 @@
+		.map (builtImage) ->
+			if not builtImage.successful
+				builtImage.error.serviceName = builtImage.serviceName
+				throw builtImage.error
+
+			d = imageDescriptorsByServiceName[builtImage.serviceName]
+			task = _.find(tasks, serviceName: builtImage.serviceName)
+
+			image =
+				serviceName: d.serviceName
+				name: d.image.tag ? d.image
+				logs: truncateString(task.logBuffer.join('\n'), LOG_LENGTH_MAX)
+				props:
+					dockerfile: builtImage.dockerfile
+					projectType: builtImage.projectType
+
+			# Times here are timestamps, so test whether they're null
+			# before creating a date out of them, as `new Date(null)`
+			# creates a date representing UNIX time 0.
+			if (startTime = builtImage.startTime)
+				image.props.startTime = new Date(startTime)
+			if (endTime = builtImage.endTime)
+				image.props.endTime = new Date(endTime)
+			docker.getImage(image.name).inspect().get('Size').then (size) ->
+				image.props.size = size
+			.return(image)
+		.tap (images) ->
+			summary = _(images).map ({ serviceName, props }) ->
+				[ serviceName, "Image size: #{humanize.filesize(props.size)}" ]
+			.fromPairs()
+			.value()
+			renderer.end(summary)
+	.finally(renderer.end)
+
+createRelease = (apiEndpoint, auth, userId, appId, composition) ->
+	_ = require('lodash')
+	crypto = require('crypto')
+	releaseMod = require('resin-release')
+	versions = require('./versions')
+
+	client = releaseMod.createClient({ apiEndpoint, auth })
+
+	releaseMod.create
+		client: client
+		user: userId
+		application: appId
+		composition: composition
+		source: "cli/#{versions.cli.major}"
+		commit: crypto.pseudoRandomBytes(16).toString('hex').toLowerCase()
+	.then ({ release, serviceImages }) ->
+		[ 'created_at', 'belongs_to__application', 'is_created_by__user', '__metadata' ].forEach (attr) ->
+			delete release[attr]
+		_.keys serviceImages, (serviceName) ->
+			img = serviceImages[serviceName]
+			[ 'created_at', 'is_a_build_of__service', '__metadata' ].forEach (attr) ->
+				delete img[attr]
+		return { client, release, serviceImages }
+
+tagServiceImages = (docker, images, serviceImages) ->
+	Promise.map images, (d) ->
+		serviceImage = serviceImages[d.serviceName]
+		imageName = serviceImage.is_stored_at__image_location
+		[ _match, registry, repo, tag = 'latest' ] = /(.*?)\/(.*?)(?::([^/]*))?$/.exec(imageName)
+		name = "#{registry}/#{repo}"
+		docker.getImage(d.name).tag({ repo: name, tag, force: true })
+		.then ->
+			docker.getImage("#{name}:#{tag}")
+		.then (localImage) ->
+			serviceName: d.serviceName
+			serviceImage: serviceImage
+			localImage: localImage
+			registry: registry
+			repo: repo
+			logs: d.logs
+			props: d.props
+
+authorizePush = (tokenAuthEndpoint, auth, registry, images) ->
+	_ = require('lodash')
+	querystring = require('querystring')
+	request = require('request')
+	getAsync = Promise.promisify(request.get)
+
+	if not _.isArray(images)
+		images = [ images ]
+
+	query = querystring.stringify
```

Comment:
NIT: you can do this using resin sdk's request like here https://github.com/resin-io/resin-preload/blob/154-multicontainer-preload/lib/preload.js#L453-L464

and avoid requiring request and querystring